### PR TITLE
only call `words_normalized` once

### PR DESCRIPTION
**Pull request by maxbachmann**
`words_normalized` should only be called once, since it is quite slow, which has a large effect now that the string matching is faster. On my laptop this achieves the following performance improvement:
Before:
```
[max@localhost dinglehopper]$ /usr/bin/time -f '\t%E real,\t%U user,\t%S sys,\t%M mmem' dinglehopper gt.txt frak2021_0.905_1587027_9141630.txt
	0:15.89 real,	9.61 user,	7.14 sys,	92704 mmem
```
After:
```
[max@localhost dinglehopper]$ /usr/bin/time -f '\t%E real,\t%U user,\t%S sys,\t%M mmem' dinglehopper gt.txt frak2021_0.905_1587027_9141630.txt
	0:12.56 real,	7.88 user,	5.56 sys,	92836 mmem
```

---
@mikegerber's task list:

- [x] Fix tests for this PR
- [x] Review error rate calculation (might also be faulty in edge cases on master. This produced NaN in test_cli_valid_json, check if it should have been infinity - according to our definition at least)
- [x] Review comment I made about the docstring 
- [x] Review RapidFuzz dependency. The one we have may be need an update for this. (tests succeed with RapidFuzz >= 3 in a fresh venv)
- [x] Review the whole PR again
- [ ] Rebase/merge into master

### CODE REVIEW

**Comment by maxbachmann** [comment]
I did now move the grapheme cluster calculation into ExtractedText, so it only has to be calculated once for each sequences. Before it was calculated twice for one of the sequences and three times for the other one. This further improves the runtime to:
```
[max@localhost dinglehopper]$ /usr/bin/time -f '\t%E real,\t%U user,\t%S sys,\t%M mmem' dinglehopper gt.txt frak2021_0.905_1587027_9141630.txt
	0:08.66 real,	6.04 user,	3.51 sys,	96860 mmem
```

So in this case we achieve a 50% speedup over the current implementation.

**Comment by maxbachmann** [comment]
I replaced uniseg with uniseg2. This is a fork of uniseg I just created, which is still Python only as well, but significantly improves the performance of all functions in uniseg. This improves the runtime of dinglehopper by another 50%
```
[max@localhost dinglehopper]$ /usr/bin/time -f '\t%E real,\t%U user,\t%S sys,\t%M mmem' dinglehopper gt.txt frak2021_0.905_1587027_9141630.txt
	0:04.20 real,	4.06 user,	0.79 sys,	89472 mmem
```

**Comment by maxbachmann** [comment]
> I replaced uniseg with uniseg2.

I am currently in contact with the uniseg author in order to get these improvements into the library.

**Comment by maxbachmann** [comment]
I opened a PR for the original library as well: https://bitbucket.org/emptypage/uniseg-py/pull-requests/1. Not sure when he will come around to review it though.

**Comment by maxbachmann** [comment]
I improved the editops implementation further and added a `score_hint`, since we already know the edit distance, which improves the runtime further to:
```
[max@localhost dinglehopper]$ /usr/bin/time -f '\t%E real,\t%U user,\t%S sys,\t%M mmem' dinglehopper gt.txt frak2021_0.905_1587027_9141630.txt
	0:02.47 real,	2.67 user,	0.71 sys,	89060 mmem
```

**Comment by maxbachmann** [comment]
My performance improvements did now get added to uniseg, so it is not required to use the forked version anymore.

@mikegerber this is now ready for review

**Comment by mikegerber** [comment]
Sorry for not reacting earlier, I had some health issues!

**Comment by maxbachmann** [comment]
> Sorry for not reacting earlier, I had some health issues!

No hurry :)
I hope your better now

**Comment by maxbachmann** [comment]
@mikegerber it appears you renamed the files used in this PR, should I reapply my changes to the new master branch?

**Comment by mikegerber** [comment]
> @mikegerber it appears you renamed the files used in this PR, should I reapply my changes to the new master branch?

Yes, we had some issues with the namespaces of our packages. You can rebase if you want or I'll do it when I make time. Sorry I didn't merge it yet, I had some (presumably small) issues with it the last time I tried (tests failing). Still want it merged!

**Comment by mikegerber** [comment]
Heads up: I'm working on merging this again.

I'm first looking into fixing running the tests using this PR, then I'm rebasing.

**Comment by mikegerber** [comment]
Note to self: Should also move on to RapidFuzz 3 if possible.

**Comment by mikegerber** [comment]
> I'm first looking into fixing running the tests using this PR, then I'm rebasing.

After cherry-picking some commits (to fix errors and warnings we already had fixed in master), only one test (test_cli_json_cer_is_infinity) fails now. Working on fixing/reviewing that one.

(Not rebased yet.)

**Comment by mikegerber** [comment]
Tests run now, but I want to review the edge cases again in the coming days. (error rate x count can be infinite or nan (if count is 0))

File: qurator/dinglehopper/extracted_text.py [hunk 1]
```diff
@@ -157,6 +179,35 @@ def text(self):
...
+    def grapheme_clusters(self):
+        if self._text is not None:
+            return self._grapheme_clusters
+        else:
+            # TODO Test with text extracted at glyph level (joiner == "")
+            clusters = []
+            for seg in self.segments:
+                clusters += seg.grapheme_clusters + self._joiner_grapheme_cluster
+            clusters = clusters[:-1]
+            return clusters
```
Comment (mikegerber) [inline qurator/dinglehopper/extracted_text.py:209]: The previous change assumed that the joiner is exactly one grapheme cluster. Well, most of the time - it can also be 0 grapheme clusters, in case we extracted at glyph level (means we glue single grapheme clusters together with nothing). Could be OK now but I need to add a test for this edge case. (Don't think anyone uses it anyway.)

**Review by mikegerber — COMMENTED** [review]
LGTM now. I've added a little check and edge case for glyph-level text.

File: qurator/dinglehopper/tests/extracted_text_test.py [hunk 2]
```diff
@@ -13,12 +13,13 @@ def test_text():
     test1 = ExtractedText(
         None,
         [
-            ExtractedText("s0", None, None, "foo"),
-            ExtractedText("s1", None, None, "bar"),
-            ExtractedText("s2", None, None, "bazinga"),
+            ExtractedText("s0", None, None, "foo", grapheme_clusters("foo")),
+            ExtractedText("s1", None, None, "bar", grapheme_clusters("bar")),
+            ExtractedText("s2", None, None, "bazinga", grapheme_clusters("bazinga")),
```
[the remark this answers is not in the document]
Comment (mikegerber) [reply qurator/dinglehopper/tests/extracted_text_test.py:18]: Ah, my idead seems to be against attrs' philosophy. I'll forget about it and think about it again when I might move on to the annotation based attrs.

File: qurator/dinglehopper/extracted_text.py (hunk 1 above)
Comment (mikegerber) [reply qurator/dinglehopper/extracted_text.py:209]: Left a TODO re: test in the code.

**Comment by mikegerber** [comment]
>     * [ ]  Review RapidFuzz dependency. The one we have may be need an update for this. (tests succeed with RapidFuzz >= 3 in a fresh venv)

2.7.0 (the minimum in this PR) runs fine, and 3.4.0 (current version) runs fine, too.

**Comment by mikegerber** [comment]
*drum rolls* now _really_ ready to merge, everything looks fine. I hope I can do it tomorro, and dive into "This branch has conflicts that must be resolved".

**Comment by maxbachmann** [comment]
> 2.7.0 (the minimum in this PR) runs fine, and 3.4.0 (current version) runs fine, too.

Yes there shouldn't be any breaking changes that would affect the usage in dinglehopper

**Comment by mikegerber** [comment]
> > 2.7.0 (the minimum in this PR) runs fine, and 3.4.0 (current version) runs fine, too.
> Yes there shouldn't be any breaking changes that would affect the usage in dinglehopper

@maxbachmann I'm being extra pedantic and check it because I've seen other software break because of a missing update of the requirements :)

**Comment by mikegerber** [comment]
Ah the uniseg dependendy needs a minimum version now, I'll add it.

**Comment by mikegerber** [comment]
> Ah the uniseg dependendy needs a minimum version now, I'll add it.

We don't need it but @maxbachmann improved the performance there too and we want the shiny speed.

(I hope the tests run after I manage to rebase this, apparently they don't here because relevant Actions config is not in this branch yet... Relevant because I only test manually on the latest Py 3.11.)

**Comment by mikegerber** [comment]
FINALLY! It is merged! Something related to PR #83 broke, but I'll fix that soon, wanted to get this in after all this time.

🍾

**Comment by mikegerber** [comment]
> Something related to PR #83 broke, but I'll fix that soon, wanted to get this in after all this time.

That was just a drained generator, I fixed it in c168155.

**Comment by maxbachmann** [comment]
Wow cool to finally see this in :partying_face:
